src/data/finsum.lean

```diff
@@ -0,0 +1,657 @@
+lemma finsum_add_distrib
+  (hf : (function.support f).finite) (hg : (function.support g).finite) :
+  ∑ᶠ i, (f + g) i = ∑ᶠ i, f i + ∑ᶠ i, g i :=
+begin
+  rw [finsum_eq_finsum_in_univ, finsum_eq_finsum_in_univ, finsum_eq_finsum_in_univ],
+  apply finsum_in_add_distrib',
+  { convert hf, simp },
+  { convert hg, simp }
+end
+
+/-- Given the function `f : α → β → M`, the finite sets `s : set α`, `t : set β`, the sum on `s`
+  over `λ x, finsum_in t (λ y, f x y)` equals the sum on `t` over `λ y, finsum_in s (λ x, f x y)`.
+  -/
+lemma finsum_in_comm {s : set α} {t : set β}
+  (f : α → β → M) (hs : s.finite) (ht : t.finite) :
+  ∑ᶠ i in s, ∑ᶠ j in t, f i j = ∑ᶠ j in t, ∑ᶠ i in s, f i j :=
+begin
+  rw [finsum_in_eq_finset_sum''' _ hs, finsum_in_eq_finset_sum''' _ ht],
+  conv_lhs { congr, skip, funext, rw finsum_in_eq_finset_sum''' _ ht },
+  conv_rhs { congr, skip, funext, rw finsum_in_eq_finset_sum''' _ hs },
+  exact finset.sum_comm,
+end
+
+/-- Given a monoid homomorphism `g : β → M`, and a function `f : α → β`, the sum on `s` over `g ∘ f`
+  equals `g` of the sum on `s` over `f`. -/
+lemma finsum_in_hom [add_comm_monoid β] (f : α → β) (g : β →+ M) (hs : s.finite) :
+  ∑ᶠ i in s, (g ∘ f) i = g (∑ᶠ j in s, f j) :=
+by rw [finsum_in_eq_finset_sum''' _ hs, finsum_in_eq_finset_sum''' _ hs,
+       set.finite.to_finset, finset.sum_hom']
+
+lemma finsum_in_hom'' [add_comm_monoid β] (f : α → β) (g : β →+ M) (hs : s.finite) :
+  ∑ᶠ i in s, g (f i) = g (∑ᶠ j in s, f j) :=
+finsum_in_hom f g hs
+
+/-- A more general version of `finsum_in_hom` that requires `s ∩ function.support f` and instead of
+  `s` to be finite. -/
+lemma finsum_in_hom' [add_comm_monoid β] {f : α → β} (g : β →+ M)
+  (h₀ : (s ∩ function.support f).finite) :
+  ∑ᶠ i in s, (g ∘ f) i = g (∑ᶠ j in s, f j) :=
+begin
+  rw [finsum_in_inter_support f, ← finsum_in_hom _ _ h₀],
+  symmetry,
+  refine finsum_in_eq_finsum_in_of_subset'' (set.inter_subset_left _ _)
+    (set.inter_subset_inter_right _ (λ x hgx, _)),
+  rw function.mem_support at hgx ⊢,
+  intro hfx, refine hgx _, simp only [hfx, function.comp],
+  exact add_monoid_hom.map_zero g
+end
+
+lemma finsum_in_hom''' [add_comm_monoid β] {f : α → β} (g : β →+ M)
+  (h₀ : (s ∩ function.support f).finite) :
+  ∑ᶠ i in s, g (f i) = g (∑ᶠ j in s, f j) :=
+finsum_in_hom' g h₀
+
+/-- Given a proposition `p` such that for all `x ∈ s, f x ≠ 0 → p x`, the sum on the elements of `s`
+  satisfied by `p` equals the sum on `s`. -/
+lemma finsum_in_inter_of_ne {p : α → Prop} (hp : ∀ x ∈ s, f x ≠ 0 → p x) :
+  ∑ᶠ i in (s ∩ p), f i = ∑ᶠ i in s, f i :=
+finsum_in_eq_finsum_in_of_subset'' (set.inter_subset_left _ _)
+  (λ _ ⟨hsx, hfx⟩, ⟨hsx, hp _ hsx $ function.mem_support.1 hfx⟩)
+
+/-- Given a proposition `p`, the sum on the elements of `s` satisfied by `p` over `f` equals the sum
+  on `s` over the function `λ a, if p a then f a else 0`. -/
+lemma finsum_in_inter (p : α → Prop) :
+  ∑ᶠ i in (s ∩ p), f i = ∑ᶠ i in s, (λ a, if p a then f a else 0) i :=
```

Comment:
```suggestion
  on `s` over the function `λ i, if p i then f i else 0`. -/
lemma finsum_in_inter (p : α → Prop) :
  ∑ᶠ i in (s ∩ p), f i = ∑ᶠ i in s, if p i then f i else 0 :=
```

Reply:
Thanks -- this came from a refactor where we added notation, in case you were wondering